doc/ch-convert_desc.rst

```diff
@@ -0,0 +1,259 @@
+
+(One could imagine a fifth format — the transient, read-only directory created
+by mounting a SquashFS archive using :code:`ch-mount(1)` — but that is not an
+image *storage* format.)
+
+The purpose of :code:`ch-convert` is to copy images, either within the same
+format or between different formats.
+
+Many of the copy algorithms have shortcuts available, which :code:`ch-convert`
+knows about, so you should typically invoke it only once per workflow, asking
+for the final format you actually need. Some copies require temporary storage
+equal to the uncompressed size of the image.
+
+
+Arguments
+=========
+
+  :code:`IN`
+    Descriptor for the input image. In the case of builder storage, this is an
+    image reference; otherwise, a filesystem path.
+
+  :code:`OUT`
+    Descriptor for the output image.
+
+  :code:`OUT_ARG ...`
+    (Valid only for :code:`tar` and :code:`squash` output formats.) Additional
+    arguments passed unchanged to the packing command.
+
+  :code:`-b`, :code:`--builder BUILDER`
+    Use image builder :code:`BUILDER`. Default: :code:`$CH_BUILDER` if set;
+    otherwise guessed (see :code:`ch-build(1)`).
+
+  :code:`-c`, :code:`--out-compress (none|gzip|xz)`
+    (Valid only for :code:`tar` and :code:`squash` output formats.)
+    Compression algorithm for archive file. Default: :code:`$CH_COMPRESSION`
+    if set; otherwise :code:`gzip`.
+
+  :code:`-h`, :code:`--help`
+    Print help and exit.
+
+  :code:`-i`, :code:`--in-fmt (builder|dir|tar|squash)`
+    Format of the input image. If omitted, inferred as described below.
+
+  :code:`--no-clobber`
+    Complain and fail if :code:`OUT` already exists, rather than replacing it.
+
+  :code:`-o`, :code:`--out-fmt (builder|dir|tar|squash)`
+    Format of the output image; inferred if omitted.
+
+  :code:`-s`, :code:`--builder-storage ARG`
+    Some builders take an argument specifying where to find their storage,
+    e.g. :code:`ch-grow --storage=/path`. If specified, pass :code:`ARG` to
+    the builder. Default: :code:`$CH_GROW_STORAGE` if set; otherwise, see
+    :code:`ch-grow(1)`. (Currently only implemented for :code:`ch-grow(1)`.)
+
+  :code:`--tmp DIR`
+    Path to temporary directory. Storage needed varies between zero and the
+    uncompressed size of the image, depending on the input and output formats.
+    Default: :code:`$TMP` if specified; otherwise :code:`/tmp`.
+
+  :code:`-v`, :code:`--verbose`
+    Print extra chatter.
+
+
+Format and filename inference
+=============================
+
+:code:`ch-convert` tries to save typing by guessing formats and filenames when
+they are reasonably clear.
+
+Format inference is done for both :code:`IN` and :code:`OUT`. The algorithm
+tries to match the value against the following globs in this order. Paths need
+not exist in the filesystem.
+
+  1. :code:`*.sqfs`, :code:`*.squash`, :code:`*.squashfs`: SquashFS.
+
+  2. :code:`*.tar`, :code:`*.t?z`, :code:`*.tar.??`, :code:`*.tar.??`: Tarball.
```

Comment:
It may be useful to try using the `file` command to infer file types. 

```
(base) $ ls
0a_directory  centos.sqfs  centos.tar.gz  centos7xz.tar.xz
(base) $ file centos.tar.gz
centos.tar.gz: gzip compressed data, last modified: Tue Mar  3 15:12:31 2020, from Unix, original size modulo 2^32 244943360 gzip compressed data, reserved method, has CRC, has comment, from FAT filesystem (MS-DOS, OS/2, NT), original size modulo 2^32 244943360
(base) $ file centos7xz.tar.xz
centos7xz.tar.xz: XZ compressed data
(base) $ file centos.sqfs
centos.sqfs: Squashfs filesystem, little endian, version 4.0, 69720732 bytes, 11087 inodes, blocksize: 131072 bytes, created: Tue Mar  3 15:12:45 2020
(base) $ file a_directory/
a_directory/: directory
```

I am not certain it works for every one of those tar formats but I would expect so. This is mostly helpful with squashfs I think. Other certain container tools using squashfs have the freedom to name it anything. Charliecloud could still work with these images if we used the text returned from `file` instead of the extension.

Reply:
One issue I see above is, how do you tell a tar-gzip from plain gzip?

For SquashFS, we could check the magic number ourselves with `dd`. I don't believe `file(1)` is POSIX, but that said I have a hard time imagining a system without it, but that said is the output format machine-parseable?

Reply:
You can ask it for all valid extensions for the file type. The bad news is that only works for gzip compressed. It might also be difficult to parse the output of file. I thought there was a flag to only print the file type, but I can't seem to make it work.